diff --git a/tests/ssl_servers.py b/tests/ssl_servers.py
index 15b071e..bbeff82 100644
--- a/tests/ssl_servers.py
+++ b/tests/ssl_servers.py
@@ -149,6 +149,8 @@ class HTTPSServerThread(threading.Thread):
 
 def make_https_server(case, *, context=None, certfile=CERTFILE,
                       host=HOST, handler_class=None):
+    import unittest
+    raise unittest.SkipTest("HTTPSServerThread Relies on select module")
     if context is None:
         context = ssl.create_default_context(ssl.Purpose.CLIENT_AUTH)
     # We assume the certfile contains both private key and certificate
