diff --git a/Makefile.am b/Makefile.am
index 9fef2be..674211b 100644
--- a/Makefile.am
+++ b/Makefile.am
@@ -2,7 +2,7 @@ AUTOMAKE_OPTIONS = foreign 1.11 no-dist-gzip dist-xz subdir-objects
 ACLOCAL_AMFLAGS = -I m4
 
 AM_CPPFLAGS = -I$(top_srcdir)/include -I$(top_builddir)/include -I.
-AM_CFLAGS = -ansi -pedantic -Wall -Wextra -Wno-long-long -Wno-unused-parameter -O2 -g
+AM_CFLAGS = -ansi -pedantic -O2 -g
 
 SUBDIRS = docs
 
